File: juno/stop_loss/basic_plus_trailing.py
from decimal import Decimal
from typing import Optional

from juno import Candle

from .basic import Basic
from .stop_loss import StopLoss
from .trailing import Trailing


class BasicPlusTrailing(StopLoss):
    _basic: Basic
    _trailing: Trailing

    def __init__(
        self,
        up_threshold: Decimal,
        up_trailing_threshold: Decimal,
        down_threshold: Optional[Decimal] = None,
        down_trailing_threshold: Optional[Decimal] = None,
    ) -> None:
        self._basic = Basic(
            up_threshold=up_threshold,
            down_threshold=down_threshold,
        )
        self._trailing = Trailing(
            up_threshold=up_trailing_threshold,
            down_threshold=down_trailing_threshold,
        )

    @property
    def upside_hit(self) -> bool:
        return self._basic.upside_hit or self._trailing.upside_hit

    @property
    def downside_hit(self) -> bool:
        return self._basic.downside_hit or self._trailing.downside_hit

    def clear(self, candle: Candle) -> None:
        self._basic.clear(candle)
        self._trailing.clear(candle)

    def update(self, candle: Candle) -> None:
        self._basic.update(candle)
        self._trailing.update(candle)

File: juno_rs/src/stop_loss/basic_plus_trailing.rs
use super::{
    Basic, BasicParams, BasicParamsContext, StopLoss, Trailing, TrailingParams,
    TrailingParamsContext,
};
use crate::{genetics::Chromosome, Candle};
use juno_derive_rs::*;
use rand::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Chromosome, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct BasicPlusTrailingParams {
    #[chromosome]
    pub basic: BasicParams,
    #[chromosome]
    pub trailing: TrailingParams,
}

pub struct BasicPlusTrailing {
    basic: Basic,
    trailing: Trailing,
}

impl BasicPlusTrailing {
    pub fn new(params: &BasicPlusTrailingParams) -> Self {
        Self {
            basic: Basic::new(&params.basic),
            trailing: Trailing::new(&params.trailing),
        }
    }
}

impl StopLoss for BasicPlusTrailing {
    fn upside_hit(&self) -> bool {
        self.basic.upside_hit() || self.trailing.upside_hit()
    }

    fn downside_hit(&self) -> bool {
        self.basic.downside_hit() || self.trailing.downside_hit()
    }

    fn clear(&mut self, candle: &Candle) {
        self.basic.clear(candle);
        self.trailing.clear(candle);
    }

    fn update(&mut self, candle: &Candle) {
        self.basic.update(candle);
        self.trailing.update(candle);
    }
}
